Make buildings explicitly provide their names as prerequisites

penev92 opened the pull request:
The `Building` trait no longer provides any custom prerequisites automatically in order to keep things consistent.
Naturally, the lint check no longer cares about buildings. Overall, this amounts to 2 bogus trait dependencies being removed.
Buildings that need to provide their name as prerequisite now have `ProvidesCustomPrerequisite` with it's default value, which is now the actor's name.
Next step is to use that last part for upgrades, so the lint check passes for #7597.

Closes #5287.

Commits will be squashed, of course.

penev92 committed:
Make buildings grant prerequisites explicitly

penev92 committed:
Add an upgrade rule for explicit prerequisite providing by buildings

penev92 committed:
Fix building prerequisites in RA

penev92 committed:
Fix building prerequisites in Cnc

penev92 committed:
Fix building prerequisites in D2k

Phrohdoh commented:
We typically use `this-notation` for trait suffixes not camelCase.

Phrohdoh commented on OpenRA.Mods.Common/Traits/Player/ProvidesCustomPrerequisite.cs:
```diff
@@ -34,12 +34,17 @@ public class ProvidesCustomPrerequisiteInfo : ITraitInfo
 	public class ProvidesCustomPrerequisite : ITechTreePrerequisite, INotifyOwnerChanged
 	{
 		readonly ProvidesCustomPrerequisiteInfo info;
+	    readonly string prerequisite;
```
Incorrect indentation.
Please run `make check` before PR-ing.

Mailaender replied:
We don't have indention checks yet. It depends on https://github.com/shuruev/StyleCopPlus/pull/5.

Phrohdoh commented on OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs:
```diff
@@ -836,6 +836,14 @@ internal static void UpgradeActorRules(int engineVersion, ref List<MiniYamlNode>
 
+				if (engineVersion < 20150330)
+				{
+					// Make buildings grant prerequisites explicitly
```
AFAIK all other comments indicating change in the upgrade rules are past-tense.

Phrohdoh commented on mods/cnc/rules/structures.yaml:
```diff
@@ -72,6 +72,7 @@ FACT:
 	WithBuildingPlacedAnimation:
 	Power:
 		Amount: 0
+	ProvidesCustomPrerequisite@buildingName:
```
You don't need these trait identifiers.

Phrohdoh replied:
You don't need trait identifiers if there is only a single instance of a trait on an actor type.

Phrohdoh replied:
The above statement is moot so it can be ignored, though I am going to ask that this becomes `@actor-name` to follow what we do in other places.

SilverlineDev commented:
Should it still be called "ProvidesCustomPrerequisites" if there are no longer any standard prerequisites?

Phrohdoh commented:
It shouldn't have ever included `Custom` (IMO) as everything is custom.
But outside of my opinion, that will need to be adjusted.

penev92 commented:
There are `ProvidesTechPrerequisite` (which needs to be renamed to `ProvidesTechLevelPrerequisite`) and `ProvidesCustomPrerequisite`, so I'm OK with "Custom".

penev92 commented:
Updated.
I still want some discussion on `ProvidesCustomPrerequisite@buildingName:` vs `ProvidesCustomPrerequisite@building-name:` vs `ProvidesCustomPrerequisite:` in this case.

pchote commented:
`ProvidesCustomPrerequisite@buildingName:` is fine.  You can't use `ProvidesCustomPrerequisite:` because that might conflict with inherited definitions.

Phrohdoh commented:
`Rename **defaul** prerequisite`

penev92 commented:
I suppose I'll wait for #7844 to be merged first to spare me some merge conflicts.

penev92 committed:
Fix building prerequisites in TS

penev92 commented:
Rebased and updated.

abcdefg30 commented:
Needs a rebase again.

penev92 commented:
Rebased and squashed.

Phrohdoh commented on OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs:
```diff
@@ -892,6 +892,14 @@ internal static void UpgradeActorRules(int engineVersion, ref List<MiniYamlNode>
+					// Made buildings grant prerequisites explicitly
+					if (depth == 0 && node.Value.Nodes.Exists(n => n.Key == "Inherits" &&
+						(n.Value.Value == "^Building" || n.Value.Value == "^BaseBuilding")))
+						node.Value.Nodes.Add(new MiniYamlNode("ProvidesCustomPrerequisite@buildingName", ""));
```
Could this becomes `@actor-name` to follow what we do in other places, please?

pchote replied:
The convention for these is mixed between`@buildingname` and `@BUILDING_NAME`.

Phrohdoh replied:
Then I would prefer `@buildingname`, or to use what I suggested as convention. But that's out of scope.

pchote replied:
I thought you were asking specifically for `ProvidesCustomPrerequisite@actor-name:`, which wasn't great.

Phrohdoh replied:
I am and see no reason not to, but if it goes against convention then it shouldn't be used.

Either way `buildingName` is bogus.

penev92 replied:
How is `buildingName` bogus and how is `buildingname` better?

Phrohdoh replied:
Because we don't use `camelCase` in yaml anywhere else that I know of.

pchote commented:
This change makes it so that `ProvidesCustomPrerequisite` is now the primary way of defining prereqs, so it is no longer Custom.  Could you please rename `ProvidesCustomPrerequisite` &rarr; `ProvidesPrerequisite`?  This may feel like scope creep, but you already have an upgrade rule with a requested change so you already need to tweak and rerun the upgrader.

Phrohdoh commented:
[I've asked for that already.](https://github.com/OpenRA/OpenRA/pull/7805#issuecomment-87893660)

penev92 commented:
I don't get why you people thing that rerunning an upgrade rule like this one on all mods is easy. It actually takes less work to do the YAML changes manually (and thrust me, it's not a small amount of work). The upgrade rule is for custom maps and mods mainly.

Also I remember a discussion about `ProvidesCustomPrerequisite` and `ProvidesTechPrerequisite` and how they should be renamed renamed. I don't remember the specifics, and the logs are gone, so we should mention that while we're on the subject.

I will wait for the new ideas about requested changes to namings to subside before going through all the YAML files manually for the N-th time.

penev92 committed:
Rename ProvidesCustomPrerequisite to ProvidesPrerequisite

penev92 commented:
Rebased again, renamed `@buildingName` to the silly-looking `@buildingname` all over the place, then renamed `ProvidesCustomPrerequisite` to `ProvidesPrerequisite`, went through all the code to see where the name needs changing (like comments and descriptions), added a new upgrade rule to change it, and finally renamed all occurances of  `ProvidesCustomPrerequisite` to `ProvidesPrerequisite` in YAML.

obrakmann commented on OpenRA.Mods.Common/Traits/Buildable.cs:
```diff
@@ -18,7 +18,7 @@ public class BuildableInfo : TraitInfo<Buildable>
 			"This can be prefixed with ! to invert the prerequisite (disabling production if the prerequisite is available)",
 			"and/or ~ to hide the actor from the production palette if the prerequisite is not available.",
 			"Prerequisites are granted by actors with the Building trait (with a prerequisite string given by the lower case actor name)",
```
> "Prerequisites are granted by actors with the Building trait (with a prerequisite string given by the lower case actor name)"

This isn't true anymore after this change. `ProvidesPrerequisite` will be the only source for prerequisites.

penev92 replied:
Along with `ProvidesTechPrerequisite`, but yeah, you're right.

obrakmann commented on OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs:
```diff
@@ -931,6 +931,18 @@ internal static void UpgradeActorRules(int engineVersion, ref List<MiniYamlNode>
+					// Made buildings grant prerequisites explicitly.
+					if (depth == 0 && node.Value.Nodes.Exists(n => n.Key == "Inherits" &&
+						(n.Value.Value == "^Building" || n.Value.Value == "^BaseBuilding")))
+						node.Value.Nodes.Add(new MiniYamlNode("ProvidesCustomPrerequisite@buildingname", ""));
```
-Custom

penev92 replied:
No, the next upgrade rule takes care of the renaming ;)

obrakmann commented on mods/d2k/rules/arrakis.yaml:
```diff
@@ -74,4 +74,6 @@ sietch:
\ No newline at end of file
+		Amount: 0
+	ProvidesPrerequisite@buildingname:
+
```
extra empty line

penev92 replied:
The utility insists that YAML files should end with 2 empty lines.

obrakmann commented:
Code looks good. I did some spot checks in-game (gdi03 in cnc and intervention in ra as both of those use custom prereqs, plus a normal skirmish in ra). Everything seemed normal. :+1: after those nits have been addressed.

penev92 commented:
Fixed the `Buildable` trait description.

pchote commented:
:+1:

pchote commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/eeda6b50dee06a042db074ee2309067d42d8850d%5E...eeda6b50dee06a042db074ee2309067d42d8850d
